src/python/pants/vcs/git.py

```diff
@@ -120,6 +164,65 @@ def __eq__(self, other: Any) -> bool:
         return id(self) == id(other)
 
 
+class ParseError(Exception):
+    pass
+
+
+class DiffParser:
+    def parse_unified_diff(self, content: str) -> FrozenDict[str, tuple[Hunk, ...]]:
+        buf = StringIO(content)
+        current_file = None
+        hunks: DefaultDict[str, list[Hunk]] = defaultdict(list)
+        for line in buf:
+            if match := self._filename_regex.match(line):
+                current_file = self._parse_filename(match)
+                continue
+
+            if match := self._lines_changed_regex.match(line):
+                if current_file is None:
+                    raise ParseError("missing filename in the diff")
+
+                try:
+                    hunk = self._parse_hunk(match, line)
+                except ValueError as e:
+                    raise ValueError(f"Failed to parse hunk: {line}") from e
+
+                hunks[current_file].append(hunk)
+                continue
+
+        return FrozenDict((filename, tuple(file_hunks)) for filename, file_hunks in hunks.items())
+
+    @cached_property
+    def _lines_changed_regex(self) -> re.Pattern:
+        return re.compile(r"^@@ -([0-9]+)(,([0-9]+))? \+([0-9]+)(,([0-9]+))? @@.*")
+
+    def _parse_hunk(self, match: re.Match, line: str) -> Hunk:
+        g = match.groups()
+        return Hunk(
+            left=TextBlock.from_count(
+                start=int(g[0]),
+                count=int(g[2]) if g[2] is not None else 1,
+            ),
+            right=TextBlock.from_count(
+                start=int(g[3]),
+                count=int(g[5]) if g[5] is not None else 1,
+            ),
+        )
+
+    @cached_property
+    def _filename_regex(self) -> re.Pattern:
+        # This only handles whitespaces. It doesn't work if a filename has something weird
+        # in it that needs escaping, e.g. a double quote.
+        return re.compile(r'^\+\+\+ ([^ ]+|"[^"]+") .*$')
```

Comment:
```suggestion
        return re.compile(r'^\+\+\+ (?:(?:"(.+)")|([^ ]+)) .*$')
```

this gives two capture groups, if the first one matches, we unescape it, the other one we can use verbatim.